### ISSUE TITLE
Wike Crashes At/Before Launch

### ISSUE THREAD

**Issue by LSeelig** [hugolabe/Wike#239]
```
user@FWLT:~$ flatpak run com.github.hugolabe.Wike 
Portal call failed: Invalid sandbox a11y own name: 'org.webkit.app-8544b3b66cebbf8d4bc96652e6245fc8bdfdf722f63bdaa4790110c167815f0e.Sandboxed.WebProcess-df4fa565-e425-4fc1-8b3c-8939301df002' doesn't match app id
MESA-INTEL: warning: ../src/intel/vulkan/anv_formats.c:993: FINISHME: support more multi-planar formats with DRM modifiers
MESA-INTEL: warning: ../src/intel/vulkan/anv_formats.c:959: FINISHME: support YUV colorspace with DRM format modifiers

** (wike:2): ERROR **: 11:52:17.315: Connection: failed to receive credentials: Expecting to read a single byte for receiving credentials but read zero bytes
```

**Comment by jntdofe** [comment]
Same problem here on GNOME OS (nightly.1020986), it might be an incompatibility with some new version of flatpak since the app hasn't been updated on flathub for 4 months and it worked before when I used it like a week ago.

**Comment by deananorth** [comment]
I'm having the same issue with Debian 13 Trixie with Gnome Wayland.

```
Portal call failed: Invalid sandbox a11y own name: 'org.webkit.app-8544b3b66cebbf8d4bc96652e6245fc8bdfdf722f63bdaa4790110c167815f0e.Sandboxed.WebProcess-0e5861d5-a44b-419d-83ed-b0fc30ca4f93' doesn't match app id

** (wike:2): ERROR **: 19:30:00.991: Connection: failed to receive credentials: Expecting to read a single byte for receiving credentials but read zero bytes
```

**Comment by ronankde** [comment]
I'm having the same issue with Debian 13 Stable with Gnome Wayland.
```
Portal call failed: Invalid sandbox a11y own name: 'org.webkit.app-8544b3b66cebbf8d4bc96652e6245fc8bdfdf722f63bdaa4790110c167815f0e.Sandboxed.WebProcess-efc00592-bc83-47a3-b036-4d4bd1d70184' doesn't match app id
MESA-INTEL: warning: ../src/intel/vulkan/anv_formats.c:993: FINISHME: support more multi-planar formats with DRM modifiers
MESA-INTEL: warning: ../src/intel/vulkan/anv_formats.c:959: FINISHME: support YUV colorspace with DRM format modifiers

** (wike:2): ERROR **: 10:26:17.497: Connection: failed to receive credentials: Era esperado ler apenas um byte para receber credenciais, mas foi lido zero byte
```

**Comment by LSeelig** [comment]
Is it still working for anyone? Is something triggering this?

**Comment by tclementdev** [comment]
Seems like it's broken for most. Could be caused by a Gnome runtime update?

**Comment by mcatanzaro** [comment]
I'm not sure what changed. [This is WebKit's fallback app ID](https://github.com/WebKit/WebKit/blob/67a085481973002fac92f712f871906d9b12f44d/Source/WTF/wtf/glib/Application.cpp#L60), used if the app does not specify one. But that hasn't changed recently. The error comes from [here](https://github.com/flatpak/flatpak/blob/10470a82d3b5ac1d710d8ca38545a9e7f0c0da1b/portal/flatpak-portal.c#L961) but this also has not changed anytime recently.

Easiest way to avoid this: call `g_set_prgname()` and set it to match your app ID. Or use GApplication.

**Comment by hugolabe** [comment]
I'm experiencing the same issue on Fedora 43.

It doesn't seem to be a problem with the application itself, which was working fine until recently.

I've tried downgrading the Flatpak package without success. Changing the GNOME runtime to version 48 also didn't work. I've done some tests with the manifest permissions.

Frankly, I'm a bit lost. I don't know if there's anything I can do to fix it from within the application itself...

**Comment by mcatanzaro** [comment]
> Easiest way to avoid this: call `g_set_prgname()` and set it to match your app ID. Or use GApplication.

Did you try this?

**Comment by mcatanzaro** [comment]
Huh, I see [you already use GApplication.](https://github.com/hugolabe/Wike/blob/8681ebdb17d7e04fdb167886bf6874ea979f3c28/src/application.py#L29) So why does `WTF::applicationID` return the fallback? There are only two possible reasons:

 * `g_application_get_default()` fails
 * `g_application_get_application_id()` fails

Both seem extremely unlikely, but I suppose it must be one or the other. Weird. I would try attaching with gdb to see what is happening.

**Comment by mcatanzaro** [comment]
Actually it's not weird, because that is the app ID used in the web process. There is no mechanism to ferry your real app ID from the UI process to the web process. So the question is: why is it only now being checked, when it was not before?

**Comment by hugolabe** [comment]
Thanks @mcatanzaro

I'm not very familiar with the inner workings of GTK and Flatpak. I understand that the WebKit process ID is being rejected, not the application ID.

Could something be achieved through Flatpak manifest permissions? I've tried `--own-name=org.webkit.app` but without success...

**Comment by tclementdev** [comment]
I seem to be able to fix the problem by reverting the Gnome 49 runtime to an older commit:

`sudo flatpak update --commit=d044cbacd9808e0f1f402885d70235218c6c5233295de860ce364949620c856c org.gnome.Platform`

**Comment by mcatanzaro** [comment]
> Could something be achieved through Flatpak manifest permissions? I've tried --own-name=org.webkit.app but without success...

No. `--own-name` is for owning D-Bus names.

**Comment by lalmeras** [comment]
The following allows me to run application (Fedora 43 Silverblue / flathub)

A working workaround :

```
cd $HOME/.var/app/com.github.hugolabe.Wike/data
flatpak run com.github.hugolabe.Wike
```

Other CWD works (like `$HOME/.var`, `$HOME/.var/app`). `$HOME/Documents`, `$HOME/.local` does not.

I do not understand how the process working directory may affect the sandboxed browser appid mismatch.

**Comment by lalmeras** [comment]
This issue seems related: https://github.com/flatpak/flatpak/issues/6053#issuecomment-2573790209

**Comment by lalmeras** [comment]
I tried to find the differences between runtimes org.fedoraproject.Platform//f43 (OK) and org.fedoraproject.Platform//f44 (KO).

With org.fedoraproject.Platform//f44, the `$HOME/.var/` working directory workaround is working.

Using dbus-monitor, I notice:
* $HOME / f43 runtime : OK, no call destination=org.freedesktop.portal.Flatpak
* $HOME/.var / f43 runtime : OK, a call to destination=org.freedesktop.portal.Flatpak is made from /usr/libexec/webkitgtk-6.0/WebKitWebProcess with sandbox-a11y-own-names=com.github.hugolabe.Wike.Sandboxed.WebProcess-bbf05fe6-da5c-4af4-b2d4-0b88d60766cf
* $HOME / f44 runtime : OK, no call destination=org.freedesktop.portal.Flatpak
* $HOME/.var / f44 runtime : KO, a call to destination=org.freedesktop.portal.Flatpak is made from /usr/libexec/webkitgtk-6.0/WebKitWebProcess with sandbox-a11y-own-names=org.webkit.app-c03c93ec8cafc307d5de617b0988d1c401bf0aae3f19e8b03174a9bc0f3cb5d7.Sandboxed.WebProcess-124945b2-3059-4e57-aaa7-eb0835cd55a2

So webkit behavior is different base on CWD. It may try to perform a11y dbus communication, but with the more recent runtime, it fails to correctly detect its appid.

The only reference to sandbox-a11y-own-names in webkit is here : https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/Launcher/glib/FlatpakLauncher.cpp#L39

It is called here : https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp#L214

The last conditional isFlatpakSpawnUsable() is here : https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp#L60

I think this call is always made (f43, f44, $HOME or $HOME/.var) as there is a call destination=org.freedesktop.portal.Flatpak with the echo command in dbus-monitor for all the use-cases. But as the result is a pointer to boolean, I cannot see what is the dbus call result. And I do not known what is the process exit code.

**Comment by lalmeras** [comment]
By using the following command switching CWD and runtime, I obtain consistent results (depends on directory, not on runtime).

```
flatpak run --command=flatpak-spawn --runtime=org.fedoraproject.Platform//f44 com.github.hugolabe.Wike --sandbox --sandbox-expose-path-ro-try=/this_path_doesnt_exist echo; echo $?
```

So this condition is the one that skip the flatpakSpawn call when $HOME/.var/.

For $HOME case, flatpakSpawn is always called, and the busName used here is broken with runtime f44 : https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/Launcher/glib/FlatpakLauncher.cpp#L91

**Comment by hugolabe** [comment]
Thanks for your tests @lalmeras 

I don't understand why this is happening, but it seems that changing the working directory solves the problem for me (Fedora 43 WS).

I've made a [pull request](https://github.com/hugolabe/Wike/pull/242) to change the CWD on app inizialization. I'm not sure if this is the best course of action, but it's currently the only option I have to fix it.

To test it, you can download the Flatpak package generated by the [action](https://github.com/hugolabe/Wike/actions/runs/24846245810) and install it manually.

**Comment by amminadabz** [comment]
That flatpak is working for me, Ultramarine (Fedora) 43

**Comment by lalmeras** [comment]
flatpak proposed by @hugolabe works for me (Fedora 43 Silverblue). Both from command line / $HOME folder and desktop launcher.

**Comment by lalmeras** [comment]
I look further into the issue. It is triggered by an early-loading of webkit performed before WikiView (WebKit.WebView) is installed into Wike application.

This early-loading triggers webkit application id discovery, but as the Adw.Application is not yet initialized, it fallbacks to a generic application id (org.webkit...). This application id is cached and will not be refreshed from my understanding (https://github.com/WebKit/WebKit/blob/527346d57f8cc02763692c6b91a8a59dfbe62413/Source/WTF/wtf/glib/Application.cpp#L39).

I succeed to fix the issue either by:
* modifying python imports to ensure that AdwApplication initialization is called before any webkit initialization (this is an ugly and intrusive fix)
* using GLib.set_prgname(...) API to setup application name early

Issue surely depends on runtime because of Webkit version (2.50.5 for f43, 2.52.1 for f44) and changes regarding applicationID discovery and/or caching.

Here is the patch for GLib.set_prgname(...).

```diff
diff --git a/src/wike.in b/src/wike.in
index 0880f17..4a63558 100755
--- a/src/wike.in
+++ b/src/wike.in
@@ -25,8 +25,9 @@ gettext.install('wike', localedir)
 
 if __name__ == '__main__':
   import gi
-  from gi.repository import Gio
+  from gi.repository import Gio, GLib
 
+  GLib.set_prgname('com.github.hugolabe.Wike')
   resource = Gio.Resource.load(os.path.join(pkgdatadir, 'wike.gresource'))
   resource._register()
 
```

**Comment by hugolabe** [comment]
Hi @lalmeras 

What you're saying makes sense, and the solution you're proposing seems excellent and much better than replacing the CWD. Thanks again for all your great work!

I will push a commit with the fix to master and will publish it in a new release shortly.

**Comment by hugolabe** [comment]
3b62ef8d45977688c21cd78b2a0019ee2b015bc5

**Comment by tclementdev** [comment]
> I look further into the issue. It is triggered by an early-loading of webkit performed before WikiView (WebKit.WebView) is installed into Wike application.

@lalmeras should this be reported to the Gnome runtime project as well? That seems like an unexpected change, what do you think?

**Comment by lalmeras** [comment]
Not sure which bugs and which projects may be notified, and the appropriate explanations needed. And I'm not fluent with flatpak/gtk/webkit techs.

# Flatpak / webkit behavior based on CWD

There's something wrong either in flatpak or webkit about the `flatpak ... --sandbox-expose-path-ro-try=/this_path_doesnt_exist`.

I don't understand what webkit is trying to do and if it is right that it succeeds or fails depending on CWD. No regression here, the behavior is the same for all versions. It just allowed to bypass the issue with the appropriate CWD.

Is this a flatpak issue or a webkit misusage ? I'm a bit lost ; I suspect something is wrong, but I don't know the flatpak or webkit expectations. And I think it is a bit hacky.

# Webkit applicationID discovery mechanism 

There's another issue with webkit applicationID discovery mechanism. But is there anything that can be done in code about it ? It may be a documentation issue or a warning message, but it may be difficult to implement based on all the use-cases addressed by webkit.

The only simple and unintrusive change would be a console message when applicationID is determined by webkit. For this issue, it would allow me to understand more quickly that webkit was initialized before Gtk application is built.

---

Do you have any insight about this analysis ? Does it seem useful ?

Good thing is this issue is well-referenced on search engine with `Portal call failed: Invalid sandbox a11y own name`, so it may help other projects.

**Comment by tclementdev** [comment]
Based on what you were saying about WebKit being initialized before Adw.Application, I'm not sure there is an issue about WebKit per se. I'm now wondering: why is WebKit initialized so early now? I don't understand it but it seems like there may be a chain of dependency that now causes this and I am wondering if that new behavior might be an unexpected change (in the Gnome runtime?).

**Comment by lalmeras** [comment]
@hugolabe I just tested Wike 3.2.1 from flathub on Fedora 44 Silverblue successfully (started from launcher or `flatpak run ...` in $HOME or $HOME/.var)

@tclementdev from my experiment, the import path was bin/wike.py -> wike.application -> wike.view (network_session). I am working on a minimal use-case.

**Comment by lalmeras** [comment]
@tclementdev I push a minimal reproducer for the issue: https://github.com/lalmeras/webkit-issue. See README for build instructions, various use-cases and explanations.

There may be other way to trigger the issue. The one I use is to perform a `WebKit.NetworkSession.get_default()` call before Gtk application initialization (in Wike, it is performed in views.py that is loaded from application.py).

**Comment by mcatanzaro** [comment]
This probably falls under the "don't do that" rule.

I don't understand why WebKit's behavior changed from one release to the next, and I don't understand why the cwd matters at all. But initializing libraries before initializing your GtkApplication is just generally not a good idea.
